Image share with new_share action and create_share action

wilsonyqm opened the pull request:
Implements the backend for #10. We will refactor the front end to use a modal and AJAX after this is merged.

rmacklin commented on app/assets/javascripts/image-sharer-modal.js:
```diff
@@ -0,0 +1,28 @@
+import $ from 'jquery';
```
Why does this file exist in this PR?

wilsonyqm replied:
Removed

rmacklin commented on app/assets/javascripts/application.js:
```diff
@@ -13,3 +13,5 @@
 //= require jquery_ujs
 //= require turbolinks
+//= require tether
+//= require bootstrap
```
These changes don't belong in this PR

wilsonyqm replied:
Deleted

rmacklin commented on app/assets/stylesheets/_submission-field.scss:
```diff
@@ -0,0 +1,3 @@
+.submission-field {
+  width: 590px; height: 390px; border: 1px solid #999999; padding: 5px;
```
Did you write this yourself?

wilsonyqm replied:
I can't remember when I write this, but I will delete it, since we are not using it now.

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -7,6 +7,7 @@ def index
     else
       @images = Image.all
     end
+    @share_forms = @images.map { |image| ShareImageForm.new(image_id: image.id) }
```
This should not be necessary

wilsonyqm replied:
Right.

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -43,3 +64,8 @@ def image_params
+
+class ShareImageForm
+  include ActiveModel::Model
+  attr_accessor :image_id, :recipient, :title
```
Where are the validations?

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -43,3 +64,8 @@ def image_params
+class ShareImageForm
+  include ActiveModel::Model
+  attr_accessor :image_id, :recipient, :title
+end
```
Please remove this.

wilsonyqm replied:
Removed

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -26,6 +27,26 @@ def create
 
+  def new_share
+    @image = Image.find(params[:id])
+    @share_form = ImagesController::ShareForm.new
```
You shouldn't need `ImagesController::` here.

wilsonyqm replied:
Deleted

rmacklin commented on app/controllers/images_controller/share_form.rb:
```diff
@@ -0,0 +1,14 @@
+  class ImagesController
```
Indentation is off here.

wilsonyqm replied:
Fixed

rmacklin commented on app/controllers/images_controller/share_form.rb:
```diff
@@ -0,0 +1,14 @@
+  class ShareForm
+    include ActiveModel::Model
+
+    attr_accessor :recipient, :title, :content
```
What made you add this `title` attribute?

rmacklin replied:
It looks like you're using it for the email subject. Let's rename it to make that clear.

rmacklin replied:
Also, do you want to provide a default for this, or make it required?

wilsonyqm replied:
I think the recipient is required here which is implemented in the validator, but the title and content can be blank.

rmacklin replied:
If the subject is not required, then you'll need to add logic for a default subject.

rmacklin commented on app/mailers/image_mailer.rb:
```diff
@@ -0,0 +1,10 @@
+class ImageMailer < ApplicationMailer
+  default from:'wilsonyqm@gmail.com'
```
This is the same as the one in ApplicationMailer, so it's unnecessary to redeclare it here.

wilsonyqm replied:
Deleted

rmacklin commented on app/views/image_mailer/send_email.html.erb:
```diff
@@ -0,0 +1,20 @@
+  <head>
+    <meta content='text/html; charset=UTF-8' http-equiv='Content-Type' />
+  </head>
+  <body>
```
This stuff is normally in a mailer layout. How come you don't have one? Did you follow the rails guide?

wilsonyqm replied:
I did follow the guide, you mean this meta html is unnecessary or the body html is not correct?

rmacklin replied:
I mean it should be in the layout, not the template itself.

rmacklin commented on app/views/image_mailer/send_email.html.erb:
```diff
@@ -0,0 +1,20 @@
+  <body>
+    <h1><%= @title %></h1>
+    <p>
+      Hi <%= @recipient %>, This is a message from wilsonyqm@gmail.com with an image<br>
```
Why are you stating your email here?

wilsonyqm replied:
Fixed

rmacklin commented on app/views/image_mailer/send_email.html.erb:
```diff
@@ -0,0 +1,20 @@
+    </p>
+    <p>The content is: <%= @content %></p>
+    <p>
+    You're welcome to check our application on:<%= ENV['APP_URL'] %>
```
Missing space after the colon.

wilsonyqm replied:
Added

rmacklin commented on app/views/image_mailer/send_email.text.erb:
```diff
@@ -0,0 +1,9 @@
+<%= @title %>
+
+Hi <%= @recipient %>, This is a message from wilsonyqm@gmail.com with an image
```
Same as above.

rmacklin commented on app/views/image_mailer/send_email.html.erb:
```diff
@@ -0,0 +1,20 @@
+    <p>
+      The image is: <%= image_tag @url %>
+    </p>
+    <p>The content is: <%= @content %></p>
```
What if they didn't submit content?

wilsonyqm replied:
Implemented in the test case

rmacklin replied:
I don't understand your response to my question.

wilsonyqm replied:
if the content is empty, the html output will be ```The content is: ``` with one more white space left on the end of string, ruby will automatically trim this whitespace if it's converted into text, so you cannot assert equal this text with send_email.text.erb now unless you implement with ```<p>The content is:<%=" #{ @content }" if @content.present? %></p>```

rmacklin commented on app/views/image_mailer/send_email.text.erb:
```diff
@@ -0,0 +1,9 @@
+
+The image is: <%= @url %>
+
+The content is: <%= @content %>
```
Same as above

rmacklin commented on app/views/image_mailer/send_email.text.erb:
```diff
@@ -0,0 +1,9 @@
+
+The content is: <%= @content %>
+
+You're welcome to check our application on https://qiaomu-image-sharer-stag.herokuapp.com
```
You should use the environment variable here.

wilsonyqm replied:
It seems like I didn't set the ENV variable, and it is not set for ENV['APP_URL'], shall I reference sendgrid setting?

rmacklin commented on app/views/images/new.html.erb:
```diff
@@ -2,6 +2,6 @@
   <%= f.input :title, label: 'Title:' %><br>
   <%= f.input :url, label: 'URL:' %><br>
-  <%= f.input :tag_list, label: 'Tags: (seperated by commas)' %><br>
+  <%= f.input :tag_list, label: 'Tags: (seperated by commas)', input_html: { value: @image.tag_list.join(', ') } %>
```
This shouldn't be part of this PR.

wilsonyqm replied:
Deleted

rmacklin commented on app/views/images/new_share.html.erb:
```diff
@@ -0,0 +1,8 @@
+<%= simple_form_for @share_form, as: :share_form, url: create_share_image_path(@image), html: { id: 'new_email_form', novalidate: '' } do |f| %>
```
Why are you customizing the id here?

wilsonyqm replied:
It is used for selection by the test cases

wilsonyqm replied:
We'll use the default id for selection in tests (`#new_share_form`)

rmacklin commented on app/views/images/new_share.html.erb:
```diff
@@ -0,0 +1,8 @@
+  <%= f.input :recipient, label: 'Recipient:', placeholder: 'user@domain.com' %><br>
+  <%= f.input :content, label: 'Content:', as: :text, placeholder: 'What you wanna tell me' %><br>
+  <%= f.label 'Image:' %><br>
+  <p><%= image_tag @image.url, class: 'image-detail__box' %></p>
```
It seems like this should be outside of the form, perhaps before it in the modal

wilsonyqm replied:
Moved it after the form.

rmacklin commented on app/views/image_mailer/send_email.html.erb:
```diff
@@ -0,0 +1,20 @@
+    <meta content='text/html; charset=UTF-8' http-equiv='Content-Type' />
+  </head>
+  <body>
+    <h1><%= @title %></h1>
```
I don't think it's necessary to repeat the email subject here.

wilsonyqm replied:
the subject is one of the input field the user wants to send, it is shown here only once.

rmacklin commented on app/views/image_mailer/send_email.text.erb:
```diff
@@ -0,0 +1,9 @@
+<%= @title %>
```
Same as above

wilsonyqm replied:
https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/31#discussion_r61339638

rmacklin commented on app/views/images/index.html.erb:
```diff
@@ -11,3 +11,7 @@
+
+<% content_for :javascript do %>
+  <%= javascript_include_tag 'index' %>
+<% end %>
```
This shouldn't be in this PR

wilsonyqm replied:
Deleted

rmacklin commented on app/views/layouts/application.html.erb:
```diff
@@ -26,5 +25,8 @@
     <%= render partial: 'layouts/footer' %>
+
+    <%= javascript_include_tag 'application' %>
+    <%= yield :javascript %>
```
Again, these changes don't belong in this PR

wilsonyqm replied:
Deleted

rmacklin commented on config/initializers/assets.rb:
```diff
@@ -8,4 +8,4 @@
 # Precompile additional assets.
 # application.js, application.css, and all non-JS/CSS in app/assets folder are already added.
-# Rails.application.config.assets.precompile += %w( search.js )
+Rails.application.config.assets.precompile += %w( index.js )
```
Should not be in this PR

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -111,4 +113,52 @@ class ImagesControllerTest < ActionController::TestCase
+
+  private
+  # method to clear the emails that have been delivered, otherwise there will be error when share image
+  # with invalid email, and this error sometimes will show up, sometimes will not.
```
Can you clarify? This seems bizarre. We should always get to the bottom of weird issues, rather than settling for "sometimes this happens, sometimes it doesn't"

wilsonyqm replied:
It works without this method in this case, but not working in the modal PR, might be some bugs in my code, that the email deliveries are not cleaned up after each test, Steven and I solved this bug with this method.

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -111,4 +113,52 @@ class ImagesControllerTest < ActionController::TestCase
+    assert_select '#new_email_form', 1
+    assert_select '.share_form_title', 1
+    assert_select '.share_form_recipient', 1
+    assert_select '.share_form_content', 1
```
I don't think we need to check for each field.

wilsonyqm replied:
Deleted

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -111,4 +113,52 @@ class ImagesControllerTest < ActionController::TestCase
+    assert_select '.share_form_title', 1
+    assert_select '.share_form_recipient', 1
+    assert_select '.share_form_content', 1
+    assert_select 'img', 1
```
Let's verify the `src` of this `img`.

wilsonyqm replied:
Added

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -111,4 +113,52 @@ class ImagesControllerTest < ActionController::TestCase
+              share_form: { title: 'some image form', recipient: 'some@jief.com', content: 'create_test' } }
+    assert ActionMailer::Base.deliveries.empty?
+    post :create_share, params
+    assert_not ActionMailer::Base.deliveries.empty?
```
Instead of this assertion and the one on line 134, consider using `assert_difference`

wilsonyqm replied:
Replaced

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -111,4 +113,52 @@ class ImagesControllerTest < ActionController::TestCase
+    assert ActionMailer::Base.deliveries.empty?
+    post :create_share, params
+    assert_not ActionMailer::Base.deliveries.empty?
+    assert_response :found
```
Please remove this. The correct assertion is on line 139.

wilsonyqm replied:
Removed

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -111,4 +113,52 @@ class ImagesControllerTest < ActionController::TestCase
+    assert_not ActionMailer::Base.deliveries.empty?
+    assert_response :found
+    assert_equal 'Shared it!', flash[:success]
+    assert_redirected_to images_path
```
You can also access ` ActionMailer::Base.deliveries.last` and make assertions about it.

wilsonyqm replied:
You mean check the email content ? it's already been done in the mailer_test.rb, and probably is unnecessary in the controller test. Right?

wilsonyqm replied:
We shoud still check something about the email so that we know the correct email was sent.

wilsonyqm replied:
Added

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -111,4 +113,52 @@ class ImagesControllerTest < ActionController::TestCase
+    assert_select '#new_email_form', 1
+    assert_select '.share_form_title', 1
+    assert_select '.share_form_recipient', 1
+    assert_select '.share_form_content', 1
```
Same as above

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -111,4 +113,52 @@ class ImagesControllerTest < ActionController::TestCase
+    assert ActionMailer::Base.deliveries.empty?
+    post :create_share, params
+    assert_response :unprocessable_entity
+    assert ActionMailer::Base.deliveries.empty?
```
Consider using `assert_no_difference`

wilsonyqm replied:
Added

rmacklin commented on test/mailers/image_mailer_test.rb:
```diff
@@ -0,0 +1,17 @@
+    assert_equal ['qiaomu@gmail.com'], email.to
+    assert_equal 'shareform', email.subject
+    assert_equal read_fixture('test_txt').join, email.text_part.body.to_s
+    assert_equal read_fixture('test_html').join, email.html_part.body.to_s
```
Let's not use fixtures. Let's just assert that the email body includes certain things that we expect it to include (in both the text part and the html part).

wilsonyqm replied:
Fixed

rmacklin commented on test/mailers/previews/image_mailer_preview.rb:
```diff
@@ -0,0 +1,4 @@
+# Preview all emails at http://localhost:3000/rails/mailers/image_mailer
+class ImageMailerPreview < ActionMailer::Preview
+
+end
```
If you aren't going to add any emails to preview, you should delete this.

wilsonyqm replied:
Deleted

rmacklin commented:
Where is the corresponding flow test?

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -111,4 +113,52 @@ class ImagesControllerTest < ActionController::TestCase
+    assert_select '.share_form_recipient', 1
+    assert_select '.share_form_content', 1
+    assert_select 'img', 1
+  end
```
What other test case is needed?

wilsonyqm replied:
test with empty content or subject?

wilsonyqm replied:
Yes

wilsonyqm replied:
Added

rmacklin commented on Gemfile:
```diff
@@ -61,3 +61,8 @@ gem 'acts-as-taggable-on', '~> 3.4'
+
+source 'https://rails-assets.org' do
+  gem 'rails-assets-tether', '>= 1.1.0'
+end
```
Lines 66-68 are also not necessary for this PR

wilsonyqm replied:
Deleted

wilsonyqm commented:
rebased by solving merge conflict

wilsonyqm commented on app/mailers/image_mailer.rb:
```diff
@@ -0,0 +1,9 @@
+class ImageMailer < ApplicationMailer
+  def send_email(image, share_form)
+    @recipient = share_form.recipient
+    @subject = share_form.subject
```
Check if the subject is `present?` - if not, provide a default subject:

```ruby
@subject = share_form.subject.presence || 'Hey check out this image'
```

wilsonyqm commented:
https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/31#discussion_r61633385

wilsonyqm commented:
Regarding https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/31#discussion_r61343639, we need to test the "share a nonexistent image" case (and handle that case)

wilsonyqm commented:
Added flow test, and address the comments, please take a look if you have time.https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/31/commits/05b820afc5f4c29077649d91c0fd55af26c4fa68

wilsonyqm commented:
Added flow test, and address the comments, please take a look if you have time.https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/31/commits/05b820afc5f4c29077649d91c0fd55af26c4fa68

rmacklin commented on app/assets/stylesheets/application.scss:
```diff
@@ -4,6 +4,9 @@ $enable-flex: true;
 @import 'utilities';
 
+@import "font-awesome-sprockets";
+@import "font-awesome";
```
Do we need both of these?

wilsonyqm replied:
Yes, we need both of them

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -27,6 +27,31 @@ def create
+      flash[:danger] = 'Image you want to share does not exist'
+      redirect_to images_path
+    end
+
```
Minor: extra newline.

wilsonyqm replied:
Deleted

rmacklin commented on app/mailers/image_mailer.rb:
```diff
@@ -0,0 +1,9 @@
+class ImageMailer < ApplicationMailer
+  def send_email(image, share_form)
+    @recipient = share_form.recipient
+    @subject = share_form.subject.presence || 'Default Subject'
```
Could we provide a more appropriate default subject than `'Default Subject'`?

wilsonyqm replied:
How about `'Oops, your subject is empty'` ?

rmacklin replied:
Well, if we want to force the user to provide a subject, we can just make it required. But if we don't I think we can come up with a sensible default, like 'Hey check out this image'

rmacklin commented on app/views/images/new_share.html.erb:
```diff
@@ -0,0 +1,11 @@
+</p>
+
+<%= simple_form_for @share_form, as: :share_form, url: create_share_image_path(@image), html: { novalidate: '' } do |f| %>
+  <%= f.input :subject, label: 'Email Subject:', placeholder: 'Hi Look At This Image' %><br>
```
Do we need `<br>`s after each input?

wilsonyqm replied:
Nope, deleted

rmacklin commented on app/views/layouts/application.html.erb:
```diff
@@ -26,5 +26,6 @@
     </div>
 
     <%= render partial: 'layouts/footer' %>
+    <%= debug(params) if Rails.env.development? %>
```
Was this change intentional?

wilsonyqm replied:
Yes. This is for debugging, which is from the tutorial book, it prints out the path, input data entered, etc.

rmacklin replied:
I agree that it can be useful for debugging, but I don't think it's necessarily worth committing the change. Also, the tutorial is only used for the authentication story, so I was just surprised to see this change here.

rmacklin commented on app/views/image_mailer/send_email.html.erb:
```diff
@@ -0,0 +1,14 @@
+  <p>
+  You're welcome to check our application on: <%= root_url %>
+  </p>
+
```
The indentation here is a bit messed up.

wilsonyqm replied:
Fixed

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -111,4 +112,49 @@ class ImagesControllerTest < ActionController::TestCase
+    end
+    assert_equal 'Shared it!', flash[:success]
+    assert_redirected_to images_path
+    assert_equal 'some image form', ActionMailer::Base.deliveries.last.subject
```
Would you like to also check that the recipient was what we expect and that the content we submitted is part of the email?

wilsonyqm replied:
Added, but I think it's been tested in the mailer_test, seems like a little bit redundant to test it again.

rmacklin replied:
That's a unit test. Testing it here is not redundant because it verifies that you've wired up the integration between the controller and the mailer properly.

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -111,4 +112,49 @@ class ImagesControllerTest < ActionController::TestCase
+      post :create_share, params
+    end
+    assert_response :unprocessable_entity
+    assert_select '#new_share_form', 1
```
We could also assert the error message we expect to see

wilsonyqm replied:
Added

rmacklin commented on test/page_objects/images/image_card.rb:
```diff
@@ -13,6 +13,11 @@ def click_tag!(tag_name)
         window.change_to(IndexPage)
       end
+
+      def click_share!
```
Let's just call this method `share!`

wilsonyqm replied:
Fixed

rmacklin commented on test/page_objects/images/share_new_page.rb:
```diff
@@ -0,0 +1,30 @@
+        node.find('img')[:src]
+      end
+
+      def image_count
```
I don't think there is much value in adding this method

wilsonyqm replied:
Deleted

rmacklin commented on test/page_objects/images/share_new_page.rb:
```diff
@@ -0,0 +1,30 @@
+        node.all('img').size
+      end
+
+      def create_share_form!(subject: nil, recipient: nil, content: nil)
```
Let's call this method `share_image!` - that's the behavior we're encapsulating.

wilsonyqm replied:
Fixed

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -94,4 +94,47 @@ class ImagesCrudTest < FlowTestCase
+    images_index_page = PageObjects::Images::IndexPage.visit
+    image_share_page = images_index_page.images[0].click_share!
+    assert_equal 1, image_share_page.image_count
+    assert_includes image_share_page.image_url, puppy_url_2
```
Why are we using `assert_includes` instead of `assert_equal` here?

wilsonyqm replied:
Fixed

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -94,4 +94,47 @@ class ImagesCrudTest < FlowTestCase
+    assert_includes image_share_page.image_url, puppy_url_2
+    images_index_page = image_share_page.create_share_form!(subject: 'subject',
+                                                         recipient: 'valid@email.com',
+                                                         content: 'some content').as_a(PageObjects::Images::IndexPage)
```
You shouldn't need to call `.as_a` here. Additionally, I'd reformat this so that there isn't so much leading whitespace. In general, I prefer:
```ruby
some_result = some_object.some_method(
  some_parameter: 1,
  some_other_parameter: 2
)
```
over:
```ruby
some_result = some_object.some_method(some_parameter: 1,
                                      some_other_parameter: 2)
```
especially when the first line is very long (like it is here)

wilsonyqm replied:
Why not use as_a here?

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -94,4 +94,47 @@ class ImagesCrudTest < FlowTestCase
+                                                         content: 'some content').as_a(PageObjects::Images::IndexPage)
+    assert_equal 'Shared it!', images_index_page.flash_message(:success)
+  end
+  test 'share image with invalid recipient' do
```
Missing newline above this. Moreover, you could include testing the invalid recipient case in the previous test, the same way that you're testing an invalid URL in the same test as a valid URL for image creation.

wilsonyqm replied:
Got you, combined these two tests together

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -94,4 +94,47 @@ class ImagesCrudTest < FlowTestCase
+    puppy_url_1 = 'http://www.pawderosa.com/images/puppies.jpg'
+    puppy_url_2 = 'http://ghk.h-cdn.co/assets/16/09/980x490/landscape-1457107485-gettyimages-512366437.jpg'
+    image_to_share = Image.create!( url: puppy_url_2, tag_list: 'superman, cute', title: 'test1' )
+    Image.create!( url: puppy_url_1, tag_list: 'cute, puppy', title: 'test2' )
```
Do we need to create two images for this test?

wilsonyqm replied:
Nope, deleted

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -94,4 +94,47 @@ class ImagesCrudTest < FlowTestCase
+    assert_includes image_share_page.image_url, puppy_url_2
+    image_to_share.destroy!
+    images_index_page = image_share_page.create_share_form!(subject: 'subject',
+                                                         recipient: 'invalid',
```
We should pass a valid email in this case. Also, the same comment about leading whitespace applies here.

wilsonyqm replied:
Fixed

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -94,4 +94,47 @@ class ImagesCrudTest < FlowTestCase
+    image_to_share.destroy!
+    images_index_page = image_share_page.create_share_form!(subject: 'subject',
+                                                         recipient: 'invalid',
+                                                         content: 'some content').as_a(PageObjects::Images::IndexPage)
```
As before, you shouldn't need to case using `as_a` here.

rmacklin commented on test/mailers/image_mailer_test.rb:
```diff
@@ -0,0 +1,42 @@
+class ImageMailerTest < ActionMailer::TestCase
+  test 'share image with valid email' do
+    image = Image.create!(title: 'share image test', url: 'http://something.com', tag_list: 'some, thing, here')
+    params = { subject: 'shareform', recipient: 'qiaomu@gmail.com', content: 'some content'}
```
Did you run pronto on this?

wilsonyqm replied:
Fixed some format problem

rmacklin commented:
Looking better. Let me know when you've addressed the new comments.

rmacklin commented:
Sending emails from your review app appears to be broken

wilsonyqm commented:
Fixed the problem of sendgrid https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/31/commits/07775c56e3868b0f377b53c2bb0ba89492c8a080, and addressed the comments https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/31/commits/1022e4dd712ef24deff81269d6ad1259f550c621

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -32,7 +32,6 @@ def new_share
     @share_form = ShareForm.new
   end
 
-  #TODO: rename to just `share` once we refactor to a modal
```
There's no need to remove this comment, because it's a TODO that you're going to do soon.

wilsonyqm replied:
Added

rmacklin commented on app/views/image_mailer/send_email.html.erb:
```diff
@@ -1,14 +1,13 @@
-  <% end %>
+    <% if @content.present? %>
+      <p>The content is: <%= @content %></p>
+    <% end %>
```
These three lines should not have been indented further. The whole file should be indented less.

wilsonyqm replied:
Fixed

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -125,32 +124,53 @@ class ImagesControllerTest < ActionController::TestCase
     assert_redirected_to images_path
     assert_equal 'some image form', ActionMailer::Base.deliveries.last.subject
+    assert_equal ['some@jief.com'], ActionMailer::Base.deliveries.last.to
+    assert_includes ActionMailer::Base.deliveries.last.text_part.body.to_s, 'create_test'
```
Consider saving ` ActionMailer::Base.deliveries.last` in a local variable and referencing that instead of repeating `ActionMailer::Base.deliveries.last`

wilsonyqm replied:
Fixed

rmacklin commented on app/controllers/images_controller/share_form.rb:
```diff
@@ -0,0 +1,14 @@
+              presence: true,
+              format: {
+                with: VALID_EMAIL,
+                message: 'Invalid email address'
```
The message should be a sentence fragment, like `'is not a valid email address'` so that `errors.full_messages` can properly construct full sentences.

wilsonyqm replied:
Make sense, does this message convention applies to all the validations for input values?

rmacklin replied:
Yes

wilsonyqm commented:
Addressed comments here https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/31/commits/abbd5be9d8c368e07c8cb6c19129b9ebd5eac503

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -94,4 +94,37 @@ class ImagesCrudTest < FlowTestCase
     assert_equal 3, images_index_page.images.count
   end
+
+  test 'share image' do
```
I would recommend moving the "share image" tests to their own file. They are not part of the CRUD operations of images.

wilsonyqm replied:
Fixed

rmacklin commented on test/mailers/image_mailer_test.rb:
```diff
@@ -0,0 +1,62 @@
+    share_form = ImagesController::ShareForm.new(params)
+    email = ImageMailer.send_email(image, share_form).deliver_now
+    assert_not ActionMailer::Base.deliveries.empty?
+    assert_equal 'Your friend sent you this image', email.subject
```
Technically you'd still want to validate that the recipient and content are correct in this case.

wilsonyqm replied:
Added

rmacklin commented on test/mailers/image_mailer_test.rb:
```diff
@@ -0,0 +1,62 @@
+    assert_not ActionMailer::Base.deliveries.empty?
+    outputs = [email.text_part.body.to_s, email.text_part.body.to_s]
+    outputs.each do |output|
+      assert_not_includes output, "The content is: #{params[:content]}"
```
Technically you'd still want to validate that the recipient and subject are correct in this case.

wilsonyqm replied:
Added

rmacklin replied:
I'm not seeing it

wilsonyqm commented:
Addressed the comments and squashed all the commits into
`- add the button and new_share form
- add the mailer
- add the create_share action
- configure sendgrid for heroku
- add flow test`

rmacklin commented on config/routes.rb:
```diff
@@ -4,8 +4,10 @@
+  resources :images, only: [:create, :new, :index, :show, :destroy] do
+    get :new_share, on: :member
+    post :create_share, on: :member
+  end
```
This change is in 18b93419d3be1546b56199f82351ea6795ddb92d but it belongs in bc3415dc8f3f513246cf42fbfcb67eda7fc14c05

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -27,6 +27,30 @@ def create
+  def new_share
+    @image = Image.find(params[:id])
+    @share_form = ShareForm.new
+  end
```
This change is in 5592d1ffcefe85cc54a8c2e3804a24927699ff92 but belongs in bc3415dc8f3f513246cf42fbfcb67eda7fc14c05

wilsonyqm replied:
Moved

rmacklin commented on test/flow_tests/share_image_test.rb:
```diff
@@ -0,0 +1,36 @@
+
+    images_index_page = PageObjects::Images::IndexPage.visit
+    image_share_page = images_index_page.images[0].share!
+    assert_includes image_share_page.image_url, puppy_url_2
```
Why is this `assert_includes` instead of `assert_equal`?

rmacklin commented on test/flow_tests/share_image_test.rb:
```diff
@@ -0,0 +1,36 @@
+    image_to_share = Image.create!( url: puppy_url_2, tag_list: 'superman, cute', title: 'test1' )
+    images_index_page = PageObjects::Images::IndexPage.visit
+    image_share_page = images_index_page.images[0].share!
+    assert_includes image_share_page.image_url, puppy_url_2
```
Same as above

rmacklin commented on test/flow_tests/share_image_test.rb:
```diff
@@ -0,0 +1,36 @@
+
+  test 'share nonexistent image' do
+    puppy_url_2 = 'http://ghk.h-cdn.co/assets/16/09/980x490/landscape-1457107485-gettyimages-512366437.jpg'
+    image_to_share = Image.create!( url: puppy_url_2, tag_list: 'superman, cute', title: 'test1' )
```
I'd remove the extra spaces around the parameters here.

wilsonyqm replied:
Fixed

rmacklin commented on test/mailers/image_mailer_test.rb:
```diff
@@ -0,0 +1,74 @@
+    assert_equal ['wilsonyqm@gmail.com'], email.from
+    assert_equal ['qiaomu@gmail.com'], email.to
+    outputs.each do |output|
+      assert_not_includes output, "The content is: #{params[:content]}"
```
https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/31#discussion_r61826130

wilsonyqm replied:
Added

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -27,6 +27,30 @@ def create
     end
   end
 
+  def new_share
```
Please move this action into the previous commit

wilsonyqm replied:
moved

wilsonyqm commented:
Fixed the commit problem, let me know if it's ready to merge

rmacklin commented on config/application.rb:
```diff
@@ -25,7 +25,7 @@ class Application < Rails::Application
     csp_settings = ["connect-src 'self'",
                     "default-src 'none'",
-                    'font-src https://cdn.jsdelivr.net',
+                    "font-src 'self' https://cdn.jsdelivr.net",
```
This belongs in the first commit because it's part of making the font-awesome changes work.

rmacklin commented:
👍 Nice job!

rmacklin commented on test/flow_tests/share_image_test.rb:
```diff
@@ -0,0 +1,36 @@
+
+    images_index_page = PageObjects::Images::IndexPage.visit
+    image_share_page = images_index_page.images[0].share!
+    assert_equal image_share_page.image_url, puppy_url_2
```
@wilsonyqm Sorry I didn't notice this this morning before we merged this PR.

The order of `assert_equal` is _expected_ then _actual_. Please fix this and other instances when you're refactoring the flow test.
